### PR TITLE
Teach publish the `--dry-run` flag

### LINKED ISSUE #1332: add --dry-run to `cargo publish`
Running `cargo package` reported no errors, but running `cargo publish` warns:

> all dependencies must come from the same source

The docs seems to indicate that `cargo package` is to be used as a "pre-flight" check of the package. I would have expected `cargo publish` to only be able to fail for reasons that involve communicating with crates.io proper.

### PR DESCRIPTION
**Pull request by JustAPerson**
Closes #1332

Open to feedback on rewording messages and such. I tried to follow [this precedent](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_generate_lockfile.rs#L48) for erroring on conflicting flags.

### CODE REVIEW

File: src/cargo/ops/registry.rs [hunk 1]
```diff
@@ -45,8 +51,13 @@ pub fn publish(manifest_path: &Path,
                                     false, true)).unwrap();
 
     // Upload said tarball to the specified destination
+    // However, do not upload if performing a dry run
     try!(config.shell().status("Uploading", pkg.package_id().to_string()));
-    try!(transmit(&pkg, &tarball, &mut registry));
+    if !dry {
```
Comment (alexcrichton) [inline src/cargo/ops/registry.rs]: Could this check be pushed into the `transmit` function actually? There's a few more errors that can happen right before the publish and it'd be nice to catch those.

**Comment by alexcrichton** [comment]
Nice! Could you also add a test for this option?

**Comment by JustAPerson** [comment]
@alexcrichton,

I've moved the check into `transmit` as requested. Additionally, I've written three tests that check the stderr/stdout of the process, but I'm not sure I can easily figure out how to check the local crates api files to ensure that the packaged crate was never uploaded properly. [There's an example of checking the local api](https://github.com/rust-lang/cargo/blob/master/tests/test_cargo_publish.rs#L64-L87) but I'm not sure what my tests should actually be looking for here. Is this step necessary?

Ignoring the issue of checking whether or not the upload actually occurred, the tests I've written cover:
* error on conflicting flags `--no-verify` and `--dry-run` 
* warn "aborting upload due to dry run" after normal output
* error when invalid readme path specified in Cargo.toml, which is checked in `transmit()`

**Comment by alexcrichton** [comment]
Those tests sound good to me! You maybe able to configure a host that's something like `http://example.com` which will never have a successful publish perhaps?

**Comment by alexcrichton** [comment]
It seems that this has stalled, so I'm going to close this, but feel free to open with a test added!

**Comment by erickt** [comment]
@JustAPerson: Any chance you could revive this? I just was thinking how wonderful this feature would be to verify a package is actually publishable before it's tagged with the release version.

**Comment by JustAPerson** [comment]
@erickt,
Sorry about that, not really sure why I gave up on this previously.
I should finally have some time this weekend after a busy semester.
I'll try to write up a few simple tests so this can be merged.

**Comment by erickt** [comment]
@JustAPerson: No worries, I know what that's like. @alexcrichton keeps closing my rust PRs because I haven't had time to get them over the finish line yet :)

**Pull request closed without merging**